### PR TITLE
Clarify documentation about the memory layout of `UnsafeCell`

### PR DESCRIPTION
**Pull request by Pointerbender**
This PR addresses a [comment](https://github.com/rust-lang/rust/pull/101717#issuecomment-1279908390) by @RalfJung in PR #101717 to further clarify the documentation of `UnsafeCell<T>`. The previous PR was merged already before we had a chance to correct this, hence this second PR :)

To goal of this PR is:

1. Split the paragraph about the memory layout of `UnsafeCell<T>` and the usage of `UnsafeCell::(raw_)get()` into two paragraphs, so that it is easier to digest for the reader.
2. Slightly simplify the previously added examples in order to reduce redundancy between the new examples and the examples that already [existed](https://github.com/rust-lang/rust/blob/ddd119b2fed57eb6b19c44c18108de95c564a48d/library/core/src/cell.rs#L1858-L1908) before these 2 PRs (which remained untouched by both PRs).

### CODE REVIEW

File: library/core/src/cell.rs [hunk 1]
```diff
@@ -1825,35 +1825,44 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 /// Therefore this is not a valid conversion, despite `NonNull<u8>` and `UnsafeCell<NonNull<u8>>>`
 /// having the same memory layout. This is because `UnsafeCell` disables niche optimizations in
 /// order to avoid its interior mutability property from spreading from `T` into the `Outer` type,
-/// thus this can cause distortions in the type size in these cases. Furthermore, it is only valid
-/// to obtain a `*mut T` pointer to the contents of a _shared_ `UnsafeCell<T>` through [`.get()`]
-/// or [`.raw_get()`]. A `&mut T` reference can be obtained by either dereferencing this pointer or
-/// by calling [`.get_mut()`] on an _exclusive_ `UnsafeCell<T>`, e.g.:
+/// thus this can cause distortions in the type size in these cases.
+///
+/// Note that it is still only valid to obtain a `*mut T` pointer to the contents of a
```
Comment (RalfJung) [inline library/core/src/cell.rs:1830]: Why "still"? In contrast to what?

File: library/core/src/cell.rs [hunk 2]
```diff
@@ -1825,35 +1825,44 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
...
+/// _shared_ `UnsafeCell<T>` through [`.get()`]  or [`.raw_get()`]. A `&mut T` reference
+/// can be obtained by either dereferencing this pointer or by calling [`.get_mut()`]
```
Comment (RalfJung) [inline library/core/src/cell.rs:1832]:
```suggestion
/// Note that the only valid way to obtain a `*mut T` pointer to the contents of a
/// _shared_ `UnsafeCell<T>` is through [`.get()`]  or [`.raw_get()`]. A `&mut T` reference
/// can be obtained by either dereferencing this pointer or by calling [`.get_mut()`]
```

**Comment by RalfJung** [comment]
Nice examples, thanks! I just left one minor edit. My other concern is that this is hidden so late in the documentation that I am a bit worried people might miss it. Are there some reasonable subsections we could use to structure this?

r? @Amanieu since this is a follow-up to https://github.com/rust-lang/rust/pull/101717

File: library/core/src/cell.rs (hunk 1 above)
Comment (Pointerbender) [reply library/core/src/cell.rs]: Thanks! I just merged your suggested edit :)

**Comment by Pointerbender** [comment]
> Are there some reasonable subsections we could use to structure this?

The new text that was added in this PR and in #101717 cover the topics of memory layout and type conversion. As the text is currently written down, I think we can combine this into one subsection labeled "Memory layout and type conversions", but an alternative could be to detangle the text even further and split it into two subsections "Memory layout" and "Type conversions". As the text is structured now, the only sentence that seems relevant to said memory layout subsection is *"`UnsafeCell<T>` has the same in-memory representation as its inner type `T`."*, the rest of the text explains about the pitfalls when converting between `UnsafeCell<T>` and `T` (namely: disabled niches, use `.get()` instead of casting). Therefore I think just adding a "Memory layout and type conversions" header feels the most natural, but please let me know what you prefer @RalfJung  :)

**Comment by RalfJung** [comment]
"Memory layout" sounds good, but "type conversion" sounds like a safe operation, like `as` or `into`.

Maybe just "memory layout"?

**Comment by Pointerbender** [comment]
> Maybe just "memory layout"?

Sounds good to me, I just updated the PR. Thanks!

**Comment by RalfJung** [comment]
LGTM, let's see what @Amanieu thinks.

**Comment by Amanieu** [comment]
LGTM!

@bors r+ rollup

**Pull request merged**